### PR TITLE
Create Wait Animation

### PR DESCRIPTION
**Pull request by eulertour**
Creates a Wait Animation which has the same effect as Scene.wait(). Also removes some of
the methods within Scene and the caching logic which is specific to waiting.

Depends on #562

### CODE REVIEW

File: manim/scene/scene.py [hunk 1]
```diff
@@ -903,6 +771,81 @@ def wait_until(self, stop_condition, max_time=60):
...
+            # Paint all non-moving objects onto the screen, so they don't
+            # have to be rendered every frame
+            moving_mobjects, static_mobjects = self.get_moving_and_static_mobjects(
+                animations
+            )
+            self.renderer.save_static_mobject_data(self, static_mobjects)
+            time_progression = self.get_animation_time_progression(animations)
+        if len(animations) == 1 and isinstance(animations[0], Wait):
+            # TODO, be smart about setting a static image
+            # the same way Scene.play does
```
Comment (huguesdevimeux) [inline manim/scene/scene.py:814]: Aren't thoses comments deprecated now? A static frame is now added at l. 796, when possible

File: manim/scene/scene.py [hunk 2]
```diff
@@ -887,6 +731,30 @@ def get_wait_time_progression(self, duration, stop_condition):
...
+        float
+            The total run_time of all of the animations in the list.
+        """
+
+        return np.max([animation.run_time for animation in animations])
+
+    def play(self, *args, **kwargs):
+        self.renderer.play(self, *args, **kwargs)
+
+    def wait(self, duration=DEFAULT_WAIT_TIME, stop_condition=None):
```
Comment (huguesdevimeux) [inline manim/scene/scene.py:755]: Suggestion : add a depreciation warning

File: manim/scene/scene.py [hunk 3]
```diff
@@ -903,6 +771,81 @@ def wait_until(self, stop_condition, max_time=60):
...
+        """
+        if len(args) == 0:
+            warnings.warn("Called Scene.play with no animations")
+            return
+
+        animations = self.compile_play_args_to_animation_list(*args, **kwargs)
+        if (
+            len(animations) == 1
+            and isinstance(animations[0], Wait)
+            and not self.should_update_mobjects()
```
Comment (huguesdevimeux) [inline manim/scene/scene.py:794]: **Question** : what will happen if there are several animations including a `Wait` and mobjects does not need to be updated ?

**Review by huguesdevimeux — APPROVED** [review]
This seems good to me.
Although there is no apparent reason it shouldn't work, did you try the caching mechanism with `Wait` with condition ?
Maybe a test for that with hashing is needed.

**Comment by eulertour** [comment]
This PR depends on #562, so we should at least get the review process started on that one before I make any changes to this one.

File: manim/scene/scene.py (hunk 2 above)
Comment (eulertour) [reply manim/scene/scene.py:755]: I don't think this should ever be deprecated, just used as more convenient syntax.

File: manim/scene/scene.py (hunk 1 above)
Comment (eulertour) [reply manim/scene/scene.py:814]: This comment is referring to instances where wait is called while some but not all Mobjects are moved by updaters. Line 796 only handles the case where wait is called and all Mobjects are stationary.

File: manim/scene/scene.py (hunk 3 above)
Comment (eulertour) [reply manim/scene/scene.py:794]: The rendered animation would still be the same, but the rendering process wouldn't benefit from this optimization. That said, that'd be a rather odd way to write a scene.

File: manim/scene/scene.py [hunk 4]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
+    ):
         """
-        This method is used to prep the animations for rendering,
-        apply the arguments and parameters required to them,
-        render them, and write them to the video file.
+        You will hardly use this when making your own animations.
+        This method is for Manim's internal use.
+
+        Returns a CommandLine ProgressBar whose fill_time
+        is dependent on the run_time of an animation,
```
Comment (behackl) [inline manim/scene/scene.py:631]:
```suggestion
        Returns a CommandLine ProgressBar whose ``fill_time``
        is dependent on the ``run_time`` of an animation,
```

File: manim/scene/scene.py [hunk 5]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
 
-        # Paint all non-moving objects onto the screen, so they don't
-        # have to be rendered every frame
-        moving_mobjects, static_mobjects = self.get_moving_and_static_mobjects(
-            animations
+        override_skip_animations: bool, optional
+            Whether or not to show skipped animations in the progress bar.
+
+        Returns
+        ------
```
Comment (behackl) [inline manim/scene/scene.py:648]:
```suggestion
        -------
```

File: manim/scene/scene.py [hunk 6]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
+    def get_animation_time_progression(self, animations):
+        """
+        You will hardly use this when making your own animations.
+        This method is for Manim's internal use.
 
-        for animation in animations:
-            animation.finish()
-            animation.clean_up_from_scene(self)
+        Uses get_time_progression to obtaina
+        CommandLine ProgressBar whose fill_time is
```
Comment (behackl) [inline manim/scene/scene.py:672]:
```suggestion
        Uses :func:`~.get_time_progression` to obtain a
        CommandLine ProgressBar whose ``fill_time`` is
```

File: manim/scene/scene.py [hunk 7]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
+        dependent on the qualities of the passed Animation,
 
-    def wait_internal(self, duration=DEFAULT_WAIT_TIME, stop_condition=None):
-        self.update_mobjects(dt=0)  # Any problems with this?
-        self.duration = duration
-        self.stop_condition = stop_condition
-        last_t = 0
+        Parameters
+        ----------
+        animations : list of Animation
```
Comment (behackl) [inline manim/scene/scene.py:677]:
```suggestion
        animations : List[:class:`~.Animation`, ...]
```

File: manim/scene/scene.py [hunk 8]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
-                if stop_condition is not None and stop_condition():
-                    time_progression.close()
-                    break
-        else:
-            self.renderer.update_frame(self)
-            dt = 1 / self.renderer.camera.frame_rate
-            self.renderer.add_frame(
-                self.renderer.get_frame(), num_frames=int(duration / dt)
+        Returns
+        ------
```
Comment (behackl) [inline manim/scene/scene.py:682]:
```suggestion
        -------
```

File: manim/scene/scene.py [hunk 9]
```diff
@@ -831,6 +731,30 @@ def get_wait_time_progression(self, duration, stop_condition):
             )
         return time_progression
 
+    def get_run_time(self, animations):
+        """
+        Gets the total run time for a list of animations.
+
+        Parameters
+        ----------
+        animations: list of Animation
```
Comment (behackl) [inline manim/scene/scene.py:740]:
```suggestion
        animations : List[:class:`Animation`, ...]
```

File: manim/scene/scene.py [hunk 10]
```diff
@@ -831,6 +731,30 @@ def get_wait_time_progression(self, duration, stop_condition):
...
+            A list of the animations whose total
+            run_time is to be calculated.
```
Comment (behackl) [inline manim/scene/scene.py:742]:
```suggestion
            ``run_time`` is to be calculated.
```

File: manim/scene/scene.py [hunk 11]
```diff
@@ -831,6 +731,30 @@ def get_wait_time_progression(self, duration, stop_condition):
...
+
+        Returns
+        ------
```
Comment (behackl) [inline manim/scene/scene.py:745]:
```suggestion
        -------
```

File: manim/scene/scene.py [hunk 12]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
         ----------
-        *args : Animation or mobject with mobject method and params
-        **kwargs : named parameters affecting what was passed in *args e.g
-            run_time, lag_ratio etc.
-        """
-        if len(args) == 0:
-            warnings.warn("Called Scene.play with no animations")
-            return
+        run_time: float
+            The run_time of the animation.
```
Comment (behackl) [inline manim/scene/scene.py:639]:
```suggestion
        run_time : float
            The ``run_time`` of the animation.
```

File: manim/scene/scene.py [hunk 13]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
 
-        animations = self.compile_play_args_to_animation_list(*args, **kwargs)
-        for animation in animations:
-            animation.begin()
+        n_iterations: int, optional
```
Comment (behackl) [inline manim/scene/scene.py:641]:
```suggestion
        n_iterations : int, optional
```

File: manim/scene/scene.py [hunk 14]
```diff
@@ -720,81 +620,81 @@ def compile_method(state):
...
+            The number of iterations in the animation.
 
-        # Paint all non-moving objects onto the screen, so they don't
-        # have to be rendered every frame
-        moving_mobjects, static_mobjects = self.get_moving_and_static_mobjects(
-            animations
+        override_skip_animations: bool, optional
```
Comment (behackl) [inline manim/scene/scene.py:644]:
```suggestion
        override_skip_animations : bool, optional
```

File: manim/scene/scene.py [hunk 15]
```diff
@@ -831,6 +731,30 @@ def get_wait_time_progression(self, duration, stop_condition):
...
+        Parameters
+        ----------
+        animations: list of Animation
+            A list of the animations whose total
+            run_time is to be calculated.
+
+        Returns
+        ------
+        float
+            The total run_time of all of the animations in the list.
```
Comment (behackl) [inline manim/scene/scene.py:747]:
```suggestion
            The total ``run_time`` of all of the animations in the list.
```

File: manim/scene/scene.py [hunk 16]
```diff
@@ -847,6 +771,81 @@ def wait_until(self, stop_condition, max_time=60):
...
+        """
+        This method is used to prep the animations for rendering,
+        apply the arguments and parameters required to them,
+        render them, and write them to the video file.
+
+        Parameters
+        ----------
+        *args : Animation or mobject with mobject method and params
+        **kwargs : named parameters affecting what was passed in *args e.g
+            run_time, lag_ratio etc.
```
Comment (behackl) [inline manim/scene/scene.py:784]: This formatting is fixed in #576, I don't add a suggestion here.

**Review by behackl — CHANGES_REQUESTED** [review]
The changes in the code look good to me. There are a few issues with the docstrings, please consider my suggestions.

**Review by leotrs — APPROVED** [review]

**Comment by leotrs** [comment]
Let's add @behackl's suggestions and fix the conflict and then merge :)

**Comment by eulertour** [comment]
This depends on #614 since it's currently based on the faulty logic for finding static mobjects.

**Comment by eulertour** [comment]
@behackl This PR is ready when you are

**Review by behackl — APPROVED** [review]
Ah, I missed 6c79273 -- good to go. 👍

**Comment by cobordism** [comment]
This PR broke the AddTextLetterByLetter animation (among others)

**Pull request merged**